### PR TITLE
#3856 hc and core refactoring

### PR DESCRIPTION
**Pull request by rustaem**
Addressing #3681.

HC and supporting parts of core refactoring

Cross-apps records properly encapsulated
Coverage
Other refactoring and fixes

### CODE REVIEW

File: apps/aecore/src/aec_chain_state.erl
Comment (cytadela8) [inline apps/aecore/src/aec_chain_state.erl:1173]: Wasn't the indentation nicer here before? Overall the indentation is better, but some reduction in indentation may be confusing.

File: apps/aecore/src/aec_block_insertion.erl
Comment (cna0) [inline apps/aecore/src/aec_block_insertion.erl:298]: How about moving this `case` to the `build_insertion_ctx_check_prev_height` clause pattern-matching?
```erlang
build_insertion_ctx_check_prev_height(key, ...) -> ...
build_insertion_ctx_check_prev_height(micro, ...) -> ...
```

File: apps/aecore/src/aec_block_insertion.erl
Comment (cna0) [inline apps/aecore/src/aec_block_insertion.erl:293]: Is it not simpler to work directly with the `#node{}` record? I think now we have to use `aec_chain_node` module everywhere introducing some more `case ... of`

Comment (rustaem) [reply apps/aecore/src/aec_block_insertion.erl:293]: > Is it not simpler to work directly with the `#node{}` record? I think now we have to use `aec_chain_node` module everywhere introducing some more `case ... of`

Sure it's easier. Like Dark Side is easier than Light Side, and eating sugar for energy is easier than exercising and sleeping well :)

File: apps/aehyperchains/src/aehc_app.erl
Comment (cna0) [inline apps/aehyperchains/src/aehc_app.erl:43]: I'm very fond of consistency in the code. Now I'm a bit confused – what should we use for logging from now on, `lager` or `aehc_log`?

**Comment by cytadela8** [comment]
I'm fine with `aec_chain_node:type` instead of accessing the record directly, but I also don't like adding those small helpers like "linfo" or `-type chain_node() :: aec_chain_node:chain_node().` I think inlining such things makes the code more readable the same applies to #3631

File: apps/aecore/src/aec_block_insertion.erl
Comment (cna0) [reply apps/aecore/src/aec_block_insertion.erl:293]: Haha. I see your point, but having a top-level module does not always reduce complexity but adds more. This could introduce a tiny bit of complexity for working with the Node record. Just an opinion; your call.

File: apps/aehyperchains/src/aehc_app.erl
Comment (cytadela8) [reply apps/aehyperchains/src/aehc_app.erl:43]: We should use lager

File: apps/aecore/src/aec_block_insertion.erl
Comment (cytadela8) [reply apps/aecore/src/aec_block_insertion.erl:293]: I think we should use `aec_chain_node:type` this is consistant with how things are done in other parts.

File: apps/aehyperchains/src/aehc_app.erl
Comment (rustaem) [reply apps/aehyperchains/src/aehc_app.erl:43]: > I'm very fond of consistency in the code. Now I'm a bit confused – what should we use for logging from now on, `lager` or `aehc_log`?

My opinion is we should abstract such things. Lager? logger? io:format? whatever?.. Those changes shouldn't be a concern for a higher level code. And abstracting is the only way to move them out of concern. Again, just my opinion here.

**Comment by cna0** [comment]
Generally, we are all fond of code readability and consistency, our opinions may only differ, Here is an opinion: I see the reason behind adding helpers for third-party libraries, but I think in the long run that might force us into having helpers for all sorts of libraries.

What did you use for code formatting? We could use it as part of the development process if the guys agree.

Overall, this pull-request looks good.

File: apps/aecore/src/aec_block_insertion.erl
Comment (rustaem) [reply apps/aecore/src/aec_block_insertion.erl:293]: > Haha. I see your point, but having a top-level module does not always reduce complexity but adds more. This could introduce a tiny bit of complexity for working with the Node record. Just an opinion; your call.

Actually, it reduces the overall complexity, and sometimes greatly so. Imagine someone from the core team add a field in `chain_node` or changes the order of fields. Even better, imagine something like that in a hot-swap scenario. It's scary, been there.

So basically we don't put records in .hrl and we definitely don't do cross-app matching. It's ok for a quick drafting sometimes, but never in a prod.

**Comment by rustaem** [comment]
> I think in the long run that might force us into having helpers for all sorts of libraries.

Which is good. Ideally, we shouldn't have more than, say, 5 lines per function, and one can only achieve that by factoring everything into most simple pieces.

> What did you use for code formatting? We could use it as part of the development process if the guys agree.

IntelliJ IDEA mostly, I really don't like rebar3_format. Not saying it's bad, it's just IDEA is so much easier for those things.

File: apps/aecore/src/aec_block_insertion.erl
Comment (cna0) [reply apps/aecore/src/aec_block_insertion.erl:293]: @rustaem good points!

File: apps/aehyperchains/src/aehc_log.erl
Comment (cytadela8) [inline apps/aehyperchains/src/aehc_log.erl:1]: This module should not be created. Those helpers only introduce confusion.

**Review by cytadela8 — COMMENTED** [review]
Please remove some helper as they make the code unnecesserly complex. aehc_log should go and inline trivial types (like `-type chain_node() :: aec_chain_node:chain_node().`)

File: apps/aehyperchains/src/aehc_consensus_hyperchains.erl
Comment (cytadela8) [inline apps/aehyperchains/src/aehc_consensus_hyperchains.erl:501]: Hey, this is cheating :P Could you maybe add a comment why this dialyzer error can't be resolved?

File: apps/aehyperchains/src/aehc_consensus_hyperchains.erl
Comment (cytadela8) [inline apps/aehyperchains/src/aehc_consensus_hyperchains.erl:600]: Here also please comment.

File: apps/aecore/src/aec_block_insertion.erl
Comment (erlmachinedev) [reply apps/aecore/src/aec_block_insertion.erl:293]: I also agree to encapsulate protocol related stuff. Function call can be useful in mocking etc..

**Review by erlmachinedev — APPROVED** [review]

**Comment by erlmachinedev** [comment]
There is a huge amount of great work! Besides that it's worth to mention few design issues which would be nice to solve in the future:

1. **aec_block_insertion.erl** was introduced as the "optimization" layer  which keeps a context of blockchain interaction. This makes some complexity (even take into account that fact that optimization impact is tremendous).  It could be nice to cover it by appropriate **behavioral API** and decouple like switchable part
2. **Mnesia bypass** functionality should be triggered to remove in the future
3. **aec_consensus behaviour** is redundant (don’t tell Gregozh about :)) and could be reduced into the more compact form
4. To organize and structure test bunches into something more manageable (at some point we discussed simulation test engine which can cover the correct resources allocation and perform garbage collection of blockchain related artifacts)

File: apps/aecore/src/aec_chain_state.erl
Comment (cytadela8) [inline apps/aecore/src/aec_chain_state.erl:350]: Consider moving this function to aec_chain_node module

File: apps/aehyperchains/include/aehc_types.hrl
Comment (cytadela8) [inline apps/aehyperchains/include/aehc_types.hrl:25]: I don't really like putting things here that have nothing todo with hc. Especially non trivial definition like signature() or query()

File: apps/aehyperchains/src/aehc_commitment.erl
Comment (cytadela8) [inline apps/aehyperchains/src/aehc_commitment.erl:41]: returns `aehc_pogf:maybe_pogf()`

File: apps/aehyperchains/src/aehc_consensus_hyperchains.erl
Comment (cytadela8) [inline apps/aehyperchains/src/aehc_consensus_hyperchains.erl:1125]: This is same case as `ldebug` helper. Remove this helper as it decreases readability.

File: apps/aehyperchains/src/aehc_consensus_hyperchains.erl
Comment (cytadela8) [inline apps/aehyperchains/src/aehc_consensus_hyperchains.erl:1117]: Remove this helper as below.

File: apps/aehyperchains/src/connectors/aehc_aeternity_connector.erl
Comment (cytadela8) [inline apps/aehyperchains/src/connectors/aehc_aeternity_connector.erl:84]: What's the situation, here?

File: rebar.config
Comment (cytadela8) [inline rebar.config:307]: OK, we can do this for now, but let's actually fix the typing in aesophia_aci_encoder later. Ok?

**Review by cytadela8 — COMMENTED** [review]
Still not a fan of removing vertical alignment, but we can live without it.

**Review by cna0 — APPROVED** [review]

File: apps/aecore/src/aec_chain_state.erl
Comment (rustaem) [reply apps/aecore/src/aec_chain_state.erl:351]: I think I've addressed most of the issues except this one. Please elaborate, why would we want to move any DB functionality into a data type module? I may be wrong, but is there any perf improvements to be achieved here, and if any, would it be worth loosing the abstraction purity?

File: rebar.config
Comment (rustaem) [reply rebar.config:307]: We'll fix it definitely, but just out of this scope.

**Review by cytadela8 — COMMENTED** [review]
Overall good, but tests are failing.

**Review by cna0 — APPROVED** [review]

**Comment by seanhinde** [comment]
Superseded by new HC design

**Pull request closed without merging**
